gunicorn/systemd.py

```diff
@@ -43,3 +44,39 @@ def listen_fds(unset_environment=True):
         os.environ.pop('LISTEN_FDS', None)
 
     return fds
+
+
+def sd_notify(state, unset_environment=False, debug=False):
+    """Send a notification to systemd. state is a string; see
+    the man page of sd_notify (http://www.freedesktop.org/software/systemd/man/sd_notify.html)
+    for a description of the allowable values.
+
+    If the unset_environment parameter is True, sd_notify() will unset
+    the $NOTIFY_SOCKET environment variable before returning (regardless of
+    whether the function call itself succeeded or not). Further calls to
+    sd_notify() will then fail, but the variable is no longer inherited by
+    child processes.
+
+    Normally this method silently ignores exceptions (for example, if the
+    systemd notification socket is not available) to allow applications to
+    function on non-systemd based systems. However, setting debug=True will
+    cause this method to raise any exceptions generated to the caller, to
+    aid in debugging."""
+
+    addr = os.getenv('NOTIFY_SOCKET')
+    if addr is None:
+        # not run in a service, just a noop
+        return
+    if unset_environment:
+        os.unsetenv('NOTIFY_SOCKET')
```

Comment:
Same as above. We could use `os.environ.pop()`.

Reply:
Also, any reason not to do this inside the `finally` block?

Reply:
I think I had a reason. but let me think about it.

Reply:
> I think I had a reason. but let me think about it.

did you find the reason?

Reply:
no special reason. I doesn't matter to me any way.